
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../../assets/images/bacteria.png">
      <meta name="generator" content="mkdocs-1.3.1, mkdocs-material-8.3.9">
    
    
      
        <title>Step 1 solution - Bioinformatics and computational approaches in Microbiology</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.1d29e8d0.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.cbb835fc.min.css">
        
          
          
          <meta name="theme-color" content="#757575">
        
      
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="grey" data-md-color-accent="red">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#solutions-step-1-taxonomic-profiling-of-metagenomic-samples-with-motus" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Bioinformatics and computational approaches in Microbiology" class="md-header__button md-logo" aria-label="Bioinformatics and computational approaches in Microbiology" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 16c0-.5-.05-.92-.12-1.32l1.57-.78-.9-1.8-1.37.69c-.55-.83-1.27-1.29-1.89-1.51l.66-1.96-1.9-.64-.76 2.28c-1.33-.13-2.12-.64-2.59-1.19l1.75-.87-.9-1.8-1.55.79a4.39 4.39 0 0 0-.72-2.02l1.55-2.32-1.66-1.11-1.41 2.12c-.48-.23-1.06-.41-1.76-.5V2H7v2.1c-.71.15-1.27.44-1.68.81L2.7 2.29 1.29 3.71l2.95 2.94C4 7.39 4 8 4 8H2v2h2.04c.06.63.17 1.36.36 2.15l-2.72.9.63 1.9 2.69-.9c.24.51.5 1.03.82 1.53l-2.38 1.59 1.11 1.66 2.52-1.68c.56.56 1.22 1.06 1.99 1.49l-.96 1.91 1.79.9 1-2-.16-.09c.95.32 2.03.54 3.27.61V22h2v-2.07c.76-.09 1.81-.29 2.77-.74l1.52 1.52 1.41-1.42-1.33-1.34c.38-.51.63-1.15.63-1.95M8.5 11A1.5 1.5 0 0 1 7 9.5 1.5 1.5 0 0 1 8.5 8 1.5 1.5 0 0 1 10 9.5 1.5 1.5 0 0 1 8.5 11m2.5 3a1 1 0 0 1-1-1 1 1 0 0 1 1-1 1 1 0 0 1 1 1 1 1 0 0 1-1 1m4.5 3a1.5 1.5 0 0 1-1.5-1.5 1.5 1.5 0 0 1 1.5-1.5 1.5 1.5 0 0 1 1.5 1.5 1.5 1.5 0 0 1-1.5 1.5Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bioinformatics and computational approaches in Microbiology
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Step 1 solution
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/sib-swiss/spring_school_bioinformatics_microbiology" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/spring_school_bioinformatics_microbiology
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Bioinformatics and computational approaches in Microbiology" class="md-nav__button md-logo" aria-label="Bioinformatics and computational approaches in Microbiology" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 16c0-.5-.05-.92-.12-1.32l1.57-.78-.9-1.8-1.37.69c-.55-.83-1.27-1.29-1.89-1.51l.66-1.96-1.9-.64-.76 2.28c-1.33-.13-2.12-.64-2.59-1.19l1.75-.87-.9-1.8-1.55.79a4.39 4.39 0 0 0-.72-2.02l1.55-2.32-1.66-1.11-1.41 2.12c-.48-.23-1.06-.41-1.76-.5V2H7v2.1c-.71.15-1.27.44-1.68.81L2.7 2.29 1.29 3.71l2.95 2.94C4 7.39 4 8 4 8H2v2h2.04c.06.63.17 1.36.36 2.15l-2.72.9.63 1.9 2.69-.9c.24.51.5 1.03.82 1.53l-2.38 1.59 1.11 1.66 2.52-1.68c.56.56 1.22 1.06 1.99 1.49l-.96 1.91 1.79.9 1-2-.16-.09c.95.32 2.03.54 3.27.61V22h2v-2.07c.76-.09 1.81-.29 2.77-.74l1.52 1.52 1.41-1.42-1.33-1.34c.38-.51.63-1.15.63-1.95M8.5 11A1.5 1.5 0 0 1 7 9.5 1.5 1.5 0 0 1 8.5 8 1.5 1.5 0 0 1 10 9.5 1.5 1.5 0 0 1 8.5 11m2.5 3a1 1 0 0 1-1-1 1 1 0 0 1 1-1 1 1 0 0 1 1 1 1 1 0 0 1-1 1m4.5 3a1.5 1.5 0 0 1-1.5-1.5 1.5 1.5 0 0 1 1.5-1.5 1.5 1.5 0 0 1 1.5 1.5 1.5 1.5 0 0 1-1.5 1.5Z"/></svg>

    </a>
    Bioinformatics and computational approaches in Microbiology
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sib-swiss/spring_school_bioinformatics_microbiology" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/spring_school_bioinformatics_microbiology
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../precourse/" class="md-nav__link">
        Precourse preparations
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          Project 1
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Project 1" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Project 1
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project1/" class="md-nav__link">
        Description
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project1/timeline/" class="md-nav__link">
        Overview
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project1/software/" class="md-nav__link">
        Software
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          Project 2
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Project 2" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Project 2
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project2/" class="md-nav__link">
        Description
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project2/Project2_Preparation/" class="md-nav__link">
        Preparation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project2/Project2_Overview/" class="md-nav__link">
        Overview
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project2/Project-2A/" class="md-nav__link">
        Project 2A
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project2/Project-2B/" class="md-nav__link">
        Project 2B
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project2/project-2C/" class="md-nav__link">
        Project 2C
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project2/Project2_Resources" class="md-nav__link">
        Resources
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          Project 3
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Project 3" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          Project 3
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        Description
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Prepare-before-start-of-the-course/" class="md-nav__link">
        Preparation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Remote-environment/" class="md-nav__link">
        Remote Environment
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Overview/" class="md-nav__link">
        Overview
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Step-1/" class="md-nav__link">
        Step 1
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Step 1 solution
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Step 1 solution
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-example-sequencing-data" class="md-nav__link">
    Download example sequencing data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#check-the-quality-of-the-sequencing-data" class="md-nav__link">
    Check the quality of the sequencing data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#filter-and-trim-reads" class="md-nav__link">
    Filter and trim reads
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxonomic-profiling-with-motus" class="md-nav__link">
    Taxonomic profiling with mOTUs
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxonomic-profiling-with-mapseq" class="md-nav__link">
    Taxonomic profiling with MAPseq
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#explore-taxonomic-profiles" class="md-nav__link">
    Explore taxonomic profiles
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Step-2/" class="md-nav__link">
        Step 2
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Step-2-sol/" class="md-nav__link">
        Step 2 solutions
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_7" type="checkbox" id="__nav_7" >
      
      
      
      
        <label class="md-nav__link" for="__nav_7">
          Project 4
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Project 4" data-md-level="1">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          Project 4
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project4/" class="md-nav__link">
        Description
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project4/project4a/" class="md-nav__link">
        Project 4A
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project4/project4b/" class="md-nav__link">
        Project 4B
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project4/software/" class="md-nav__link">
        Software
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../project4/timeline/" class="md-nav__link">
        Project timeline
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-example-sequencing-data" class="md-nav__link">
    Download example sequencing data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#check-the-quality-of-the-sequencing-data" class="md-nav__link">
    Check the quality of the sequencing data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#filter-and-trim-reads" class="md-nav__link">
    Filter and trim reads
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxonomic-profiling-with-motus" class="md-nav__link">
    Taxonomic profiling with mOTUs
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#taxonomic-profiling-with-mapseq" class="md-nav__link">
    Taxonomic profiling with MAPseq
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#explore-taxonomic-profiles" class="md-nav__link">
    Explore taxonomic profiles
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
  <a href="https://github.com/sib-swiss/spring_school_bioinformatics_microbiology/edit/master/docs/project3/Step-1-sol.md" title="Edit this page" class="md-content__button md-icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>



<h1 id="solutions-step-1-taxonomic-profiling-of-metagenomic-samples-with-motus">Solutions Step 1: Taxonomic profiling of metagenomic samples with mOTUs</h1>
<p>General note: this guide has been written assuming you use a Mac or Linux Command Line.</p>
<hr />
<h2 id="download-example-sequencing-data">Download example sequencing data</h2>
<p>Explore the files, in particular you can check:</p>
<ul>
<li>
<p>How many reads there are per sample?
    <details>
    <summary markdown="span">Solution 1</summary></p>
<p>Knowing that each read takes up four lines in the fastq file, we can simply count the number of lines with <code>wc -l</code> and divide the result by <code>4</code>. The following command does it all in one line. 
<div class="highlight"><pre><span></span><code><span class="nb">echo</span> <span class="k">$(</span>cat sampleA_1.fastq<span class="p">|</span>wc -l<span class="k">)</span>/4<span class="p">|</span>bc
</code></pre></div>
</details> </p>
<p><details>
<summary markdown="span">Solution 2</summary></p>
<p>We can count the number of lines with <code>@read</code>:
<div class="highlight"><pre><span></span><code>grep -c <span class="s2">&quot;@read&quot;</span> sampleA_1.fastq
</code></pre></div></p>
</details>
</li>
<li>
<p>What is the average length of the reads? Is there a difference between the read lengths in the forward and reverse files?
    <details>
    <summary markdown="span">Solution 1</summary></p>
<p>We can first extract only the sequences:
<div class="highlight"><pre><span></span><code>grep -A <span class="m">1</span> <span class="s2">&quot;@read&quot;</span> sampleA_1.fastq  <span class="p">|</span> grep -v <span class="s2">&quot;\-\-&quot;</span> <span class="p">|</span> grep -v <span class="s2">&quot;read&quot;</span> &gt; sequences_sampleA_1
</code></pre></div>
With <code>-A 1</code> we select also 1 row after the match. With <code>grep -v</code> we remove what is not needed. We can now check the length with:
<div class="highlight"><pre><span></span><code>cat sequences_sampleA_1 <span class="p">|</span> awk <span class="s1">&#39;{print length}&#39;</span>
</code></pre></div>
This will print a big list, we can count how many times each length appear:
<div class="highlight"><pre><span></span><code>cat sequences_sampleA_1 <span class="p">|</span> awk <span class="s1">&#39;{print length}&#39;</span> <span class="p">|</span> sort <span class="p">|</span> uniq -c <span class="p">|</span> sort -n <span class="p">|</span> tail
</code></pre></div>
Which produces:
<div class="highlight"><pre><span></span><code>  <span class="m">238</span> <span class="m">96</span>
  <span class="m">242</span> <span class="m">92</span>
  <span class="m">258</span> <span class="m">93</span>
  <span class="m">341</span> <span class="m">97</span>
  <span class="m">344</span> <span class="m">94</span>
  <span class="m">346</span> <span class="m">95</span>   
  <span class="m">428</span> <span class="m">98</span>
  <span class="m">849</span> <span class="m">20</span>
 <span class="m">1194</span> <span class="m">99</span>
<span class="m">61306</span> <span class="m">100</span>
</code></pre></div>
So the majority of the reads have length 100 (61,306 out of 67,926, 90%)</p>
</details>
<p><details>
<summary markdown="span">Solution 2</summary></p>
<p>To quickly check the average length of the reads in the terminal, do: 
<div class="highlight"><pre><span></span><code>awk <span class="s1">&#39;NR%4==2{sum+=length($0)}END{print sum/(NR/4)}&#39;</span> sampleA_1.fastq
</code></pre></div>
The average read length in the reverse reads seem to be slightly lower in all the samples.</p>
</details>
</li>
<li>
<p>Do you have the same read IDs in the forward and reverse file? 
    <details>
    <summary markdown="span">Solution</summary></p>
<p>Since these are paired reads, the read ids should be identical and in the same order. You can check this in the terminal like so: </p>
<div class="highlight"><pre><span></span><code><span class="c1">#get list of read ids from the forward and reverse files</span>
grep <span class="s1">&#39;@read&#39;</span> sampleA_1.fastq &gt; sampleA_ids_1.txt
grep <span class="s1">&#39;@read&#39;</span> sampleA_2.fastq &gt; sampleA_ids_2.txt
<span class="c1">#check if they are identical </span>
diff -s sampleA_ids_1.txt  sampleA_ids_2.txt
</code></pre></div>
</details>
</li>
</ul>
<p>There are many different ways of performing the same task. If you have done something different and accomplished the same thing, awesome!</p>
<h2 id="check-the-quality-of-the-sequencing-data">Check the quality of the sequencing data</h2>
<ul>
<li>
<p>Which part of the reads is of lower quality?
    <details>
    <summary markdown="span">Solution</summary></p>
<p>The ends of the reads are typically of lower quality. This is to be expected. The quality of calls typically degrades as the run progresses due to problems in the sequencing chemistry. </p>
<p></details>
- Is there any difference between the quality of the forward and reverse reads?
<details>
<summary markdown="span">Solution</summary></p>
<p>Reverse reads are usually of lower quality than forward reads, particularly at the read ends. Again this is due to the way paired end sequencing is performed with the forward orientiation is sequenced first followed by the reverse orientation. </p>
</details>
</li>
</ul>
<h2 id="filter-and-trim-reads">Filter and trim reads</h2>
<ul>
<li>
<p>Try to run trimmomatic (you can use different parameters).</p>
<p><details>
<summary markdown="span">Solution</summary></p>
<p>Here is an example command: </p>
<p><div class="highlight"><pre><span></span><code>trimmomatic PE sampleA_1.fastq sampleA_2.fastq sampleA_filtered_1P.fastq sampleA_filtered_1U.fastq sampleA_filtered_2P.fastq sampleA_filtered_2U.fastq ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36
</code></pre></div>
Here is a description of the parameters used in this specific command. You may have explored other parameters too. </p>
<ul>
<li><code>ILLUMINACLIP</code>: specifies the file containing the adapter sequences to trim and additional parameters on how to perform the adapter trimming. <code>TruSeq3-PE.fa</code> (provided by <code>trimmomatic</code>) contains the Illumina adapter sequences used by HiSeq and MiSeq machines. </li>
<li><code>LEADING</code>: Remove low quality bases (lower than <code>3</code>) from the beginning of the reads. </li>
<li><code>TRAILING</code>: Remove low quality bases (lower than <code>3</code>) from the ends of the reads.</li>
<li><code>SLIDINGWINDOW</code>: consider a window of bases (here <code>4</code> at once) and trim once the average quality within the window falls below a threshold quality (here <code>15</code>). </li>
<li><code>MINLEN</code>: remove reads lower than the specified min length (here <code>36</code>)
</details></li>
</ul>
</li>
<li>
<p>How many files did trimmomatic generate? What do they contain?
    <details>
    <summary markdown="span">Solution</summary></p>
<p>4 files are produced</p>
<ul>
<li>sampleA_filtered_1P, containing the forward reads that pass the filter and have a mate (in filtered_2P);</li>
<li>sampleA_filtered_1U, containing the forward reads that pass the filter and do not have a mate (the paired reverse read didn&rsquo;t pass the filter)</li>
<li>sampleA_filtered_2P, containing the reverse reads that pass the filter and have a mate (in filtered_1P);</li>
<li>sampleA_filtered_2U, containing the reverse reads that pass the filter and do not have a mate (the paired forward read didn&rsquo;t pass the filter)</li>
</ul>
<p></details>
- How many reads have been filtered out?
<details>
<summary markdown="span">Solution</summary></p>
<p>866 reads (1.27%) of all reads were filtered out from sampleA using the above parameters. </p>
</details>
</li>
<li>
<p>Check the quality of the filtered reads. Did the quality improve?
    <details>
    <summary markdown="span">Solution</summary></p>
<p>You can check the quality of the filterd reads again with fastqc. </p>
<p><div class="highlight"><pre><span></span><code>fastqc sampleA_filtered_1P.fastq
fastqc sampleA_filtered_2P.fastq
</code></pre></div>
 The quality of reads (particularly of the reverse reads) has improved!
</details></p>
</li>
</ul>
<h2 id="taxonomic-profiling-with-motus">Taxonomic profiling with mOTUs</h2>
<ul>
<li>
<p>Use <code>motus</code> (manual: <a href="https://github.com/motu-tool/mOTUs_v2#simple-examples">link</a>) to create a profile from the files created by trimmomatic.
    <details>
    <summary markdown="span">Solution</summary>
     Here is the mOTU command to generate a taxonomic profile using default parameters.</p>
<div class="highlight"><pre><span></span><code>motus profile -f sampleA_filterd_1P.fastq -r sampleA_filtered_2P.fastq -s sampleA_filtered_1U.fastq,sampleA_filtered_2U.fastq -n sampleA -o sampleA_profile.txt
</code></pre></div>
<p></details>
- How many species are detected? How many are reference species and how many are unknown species?
<details>
<summary markdown="span">Solution</summary>
You can quickly check how many species were detected with:</p>
<p><div class="highlight"><pre><span></span><code><span class="c1">#this also counts unassigned so subtract 1 from the result</span>
grep -c -v <span class="s1">&#39;0.0000000000\|#&#39;</span> sampleA_profile.txt
</code></pre></div>
<code>97</code> species were dectected. 
Around <code>3.4 %</code> were unassigned . </p>
<p>You can check how many ref-mOTUs were detected using these command:
<div class="highlight"><pre><span></span><code>grep -v <span class="s1">&#39;0.0000000000\|#&#39;</span> sampleA_profile.txt &gt; sampleA_profile_detected.txt
grep -c <span class="s1">&#39;ref_mOTU_v3_&#39;</span> sampleA_profile_detected.txt
</code></pre></div>
<code>39</code> ref-mOTUs were detected in sampleA. Note that this number is also reported as stdout when you run <code>motus profile</code></p>
</details>
</li>
<li>
<p>Can you change some parameters in <code>motus</code> to profile more or less species? (Hint, look <a href="https://github.com/motu-tool/mOTUs/wiki/Increase-precision-or-recall">here</a>)
        <details>
    <summary markdown="span">Solution</summary>
    Precision is the number of TP out of the total number of detected species. </p>
<p>Recall is the number of detected species out of all the species actually present in the sample. </p>
<p>To increase precision at the cost of recall you can increase parameters <code>-g</code> (default: 3) and -l (default: 75). </p>
<p><div class="highlight"><pre><span></span><code>motus profile -f sampleA_filterd_1P.fastq -r sampleA_filtered_2P.fastq -s sampleA_filtered_1U.fastq,sampleA_filtered_2U.fastq -n sampleA -o sampleA_profile_high_p.txt -g <span class="m">8</span> -l <span class="m">90</span>
</code></pre></div>
We have detected just 37 species. </p>
<p>To increase recall at the cost of having more false positives you can do:</p>
<p><div class="highlight"><pre><span></span><code>motus profile -f sampleA_filterd_1P.fastq -r sampleA_filtered_2P.fastq -s sampleA_filtered_1U.fastq,sampleA_filtered_2U.fastq -n sampleA -o sampleA_profile_high_r.txt -g <span class="m">1</span> -l <span class="m">45</span>
</code></pre></div>
We have detected <code>331</code> species. </p>
</details>
</li>
<li>
<p>How can you merge different motus profiles into one file? Try to profile and then merge three profiles (Sample A, B and C).</p>
<p><details>
<summary markdown="span">Solution</summary></p>
<p>After creating the individual tax profiles for all the samples, do: </p>
<p><div class="highlight"><pre><span></span><code>motus merge - i sampleA_profile.txt,sampleB_profile.txt,sampleC_profile.txt -o merged_profiles.txt
</code></pre></div>
This results in a tab-separated file containing the tax profiles. 
</details></p>
</li>
</ul>
<h2 id="taxonomic-profiling-with-mapseq">Taxonomic profiling with MAPseq</h2>
<ul>
<li>
<p>Similar as with mOTUs, first create a profile for each sample (A,B, and C) and then merge them into one (Check the <a href="https://github.com/jfmrod/MAPseq">github page</a> for the command). </p>
<p><details>
<summary markdown="span">Solution</summary></p>
<p>In order to create a taxonomic profile using <code>MAPseq</code> for sampleA do:</p>
<p><div class="highlight"><pre><span></span><code>mapseq sampleA_filtered_1P.fasta &gt; sampleA.mseq
</code></pre></div>
<code>MAPseq</code> seems to be a bit faster than mOTUs (took ~2 min to run)</p>
<p><code>sampleA.mseq</code> contains the results from mapping reads to the reference database of OTUs provided by <code>MAPseq</code> (alignment score, database hit, etc) and the taxnomic classifications along with associated confidences. </p>
<p>After generating the <code>.mseq</code> files for all the samples, you can merge them into one OTU table like so:</p>
<p><div class="highlight"><pre><span></span><code>mapseq -otutable sampleA.mseq sampleB.mseq sampleC.mseq -ti <span class="m">1</span> -tl <span class="m">3</span> &gt; mapseq_otutable_otu97.tsv
</code></pre></div>
This creates an OTU table containing reads mapped to 97% OTUs for sampleA, sampleB and sampleC. </p>
<p>Note that depending on whether <code>-ti</code> is 0 or 1, what <code>-tl</code> means changes. </p>
<p>If you have <code>-ti 0</code>, then <code>-tl</code> indicates the taxonomic level (0 (domain), 1 (phylum), 2 (class), 3 (order), 4 (family), 5 (genus), 6 (species)) . So if <code>-ti 0 -tl 3</code> means that the OTU table will report only read counts mapping to order-level NCBI taxonomies. </p>
<p>If you have <code>-ti 1</code>, then <code>-tl</code> indicates the OTU clustering level (1 (90% OTU), 2 (96% OTU), 3 (97% OTU), 4 (98%), 5 (99%)) . So if <code>-ti 1 -tl 3</code> means that the OTU table will report only read counts mapping to 97% OTUs. </p>
<p>To obtain reads mapping to 99% OTUs : 
<div class="highlight"><pre><span></span><code>mapseq -otutable sampleA.mseq sampleB.mseq sampleC.mseq -ti <span class="m">0</span> -tl <span class="m">5</span> &gt; mapseq_otutable_otu99.tsv
</code></pre></div></p>
<p>If we increase the clustering level to 99%, we observe fewer species detected for all the samples. This might be because at a finer resolution, we might not be able to assign taxonomies too well resulting in a smaller number of species being profiled. </p>
<p>To obtain reads mapping to 96% OTUs : </p>
<div class="highlight"><pre><span></span><code>mapseq -otutable sampleA.mseq sampleB.mseq sampleC.mseq -ti <span class="m">0</span> -tl <span class="m">2</span> &gt; mapseq_otutable_otu96.tsv
</code></pre></div>
<p>If we decrease the clustering level to 96%, we observe more species detected for all the samples. </p>
</details>
</li>
<li>
<p>Can you compare mOTUs and MAPseq profiles?
    <details>
    <summary markdown="span">Solution</summary></p>
<table>
<thead>
<tr>
<th>Profiler</th>
<th>species detected in sampleA</th>
</tr>
</thead>
<tbody>
<tr>
<td>mOTUs</td>
<td>97</td>
</tr>
<tr>
<td>MAPseq 97%</td>
<td>173</td>
</tr>
<tr>
<td>MAPseq 99%</td>
<td>121</td>
</tr>
<tr>
<td>MAPseq 96%</td>
<td>179</td>
</tr>
</tbody>
</table>
<p>It looks like mOTUs is a bit more conservative at detecting species. Note that since mOTUs and OTUs are defined differently, it might not be straightforward to make a direct comparison. 
</details></p>
</li>
</ul>
<h2 id="explore-taxonomic-profiles">Explore taxonomic profiles</h2>
<p>Explore the taxonomic profiles (<code>tax_profile</code>).  </p>
<details>
<summary markdown="span">Solution</summary>

Load packages and taxonomic profile into your R environment

<div class="highlight"><pre><span></span><code><span class="c1">#packages</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>

<span class="c1">#tax profiles</span>
<span class="nf">load</span><span class="p">(</span><span class="nf">url</span><span class="p">(</span><span class="s">&quot;https://zenodo.org/record/6517497/files/human_microbiome_dataset.Rdata&quot;</span><span class="p">))</span>
</code></pre></div>

Let's have a quick peek at our data. 
<div class="highlight"><pre><span></span><code><span class="nf">dim</span><span class="p">(</span><span class="n">tax_profile</span><span class="p">)</span>
</code></pre></div>

There are 246 rows and 496 columns. Let's check the content:

<div class="highlight"><pre><span></span><code><span class="n">tax_profile</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">]</span>
</code></pre></div>
Which results in:
<div class="highlight"><pre><span></span><code>             700002_T0 700002_T1 700002_T2
Blautia            <span class="m">229</span>      <span class="m">1196</span>      <span class="m">1758</span>
Bacteroides       <span class="m">4018</span>      <span class="m">1705</span>      <span class="m">1660</span>
Agathobacter        <span class="m">80</span>      <span class="m">1580</span>       <span class="m">126</span>
</code></pre></div>
The rows are genera and the columns are samples.

<div class="highlight"><pre><span></span><code><span class="nf">head</span><span class="p">(</span><span class="n">metadata</span><span class="p">)</span>
</code></pre></div>

Which is:
<div class="highlight"><pre><span></span><code>          Subject Timepoint Sex
700002_T0  <span class="m">700002</span>         <span class="m">0</span>   M
700002_T1  <span class="m">700002</span>        <span class="m">12</span>   M
700002_T2  <span class="m">700002</span>        <span class="m">24</span>   M
700002_T3  <span class="m">700002</span>        <span class="m">50</span>   M
700004_T0  <span class="m">700004</span>         <span class="m">0</span>   F
700004_T1  <span class="m">700004</span>        <span class="m">12</span>   F
</code></pre></div>

The 496 samples are from 124 patients (sampled 4 times over the course of the year). `tax_profile` contains the read counts of of species (rows) across all samples (columns).

For each sample, we know the corresponding subject id, timepoint of sampling and the sex of the subject. 


</details>

<p>Here are some hints of what you can check:</p>
<ul>
<li>
<p>How many reads there are per sample?
    <details>
    <summary markdown="span">Solution</summary></p>
<p>Calculate total number of reads in every sample (also called library size of sample)</p>
<div class="highlight"><pre><span></span><code><span class="n">sample_read_counts</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">total_read_counts</span> <span class="o">=</span> <span class="nf">colSums</span><span class="p">(</span><span class="n">tax_profile</span><span class="p">))</span>
</code></pre></div>
<p>How are the sample read counts distributed? 
<div class="highlight"><pre><span></span><code><span class="nf">ggplot</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">sample_read_counts</span><span class="p">)</span> <span class="o">+</span> <span class="nf">geom_histogram</span><span class="p">(</span><span class="n">mapping</span> <span class="o">=</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">total_read_counts</span><span class="p">),</span> <span class="n">bins</span> <span class="o">=</span> <span class="m">60</span><span class="p">)</span> <span class="o">+</span> <span class="nf">ylab</span><span class="p">(</span><span class="s">&#39;Number of samples&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">xlab</span><span class="p">(</span><span class="s">&#39;Total number of reads&#39;</span><span class="p">)</span>
</code></pre></div>
Which results in:</p>
<p><img alt="" src="../../assets/images/Project3/step1_hist_n_read.png" /></p>
<p>Note how variable the total read counts are across all samples. This is a problem because this variation is most likely a result of the sequencing process and not any meaningful biological variation.</p>
</details>
</li>
<li>
<p>If you want to compare different samples, is it a problem that there are different read counts? Try to divide each value within a sample by the sum of the reads in that sample to normalise the data (also called relative abundance).</p>
<p><details>
<summary markdown="span">Solution</summary></p>
<p>We need to remove this technical variation caused by differing total read counts to meaningfully compare samples. 
We can do this by <strong>relative abundance normalization</strong> where we divide each value within a sample by the total read counts in that sample.</p>
<div class="highlight"><pre><span></span><code><span class="n">rel_ab</span> <span class="o">=</span> <span class="nf">prop.table</span><span class="p">(</span><span class="n">tax_profile</span><span class="p">,</span><span class="m">2</span><span class="p">)</span>
</code></pre></div>
<p>Now the abundances of each sample should sum to 1. </p>
<div class="highlight"><pre><span></span><code><span class="nf">all</span><span class="p">(</span><span class="nf">colSums</span><span class="p">(</span><span class="n">rel_ab</span><span class="p">)</span> <span class="o">==</span> <span class="m">1</span><span class="p">)</span>
</code></pre></div>
</details>
</li>
<li>
<p>Which genera are the most and least prevalent?</p>
<p><details>
<summary markdown="span">Solution</summary></p>
<p>Prevalence of a species refers to the the proportion of samples in which that species is detected. </p>
<p>First we calculate prevalence for all the genera. 
<div class="highlight"><pre><span></span><code><span class="n">number_of_samples</span> <span class="o">=</span> <span class="nf">dim</span><span class="p">(</span><span class="n">tax_profile</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span>
<span class="n">prevalence_df</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">Prevalence</span> <span class="o">=</span> <span class="nf">rowSums</span><span class="p">(</span><span class="n">rel_ab</span><span class="o">&gt;</span><span class="m">0</span><span class="p">)</span><span class="o">/</span><span class="n">number_of_samples</span><span class="p">,</span> <span class="n">genus</span> <span class="o">=</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">tax_profile</span><span class="p">))</span>
</code></pre></div></p>
<p>You can plot an histogram of the prevalences:
<div class="highlight"><pre><span></span><code><span class="nf">ggplot</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">prevalence_df</span><span class="p">)</span> <span class="o">+</span> <span class="nf">geom_histogram</span><span class="p">(</span><span class="n">mapping</span> <span class="o">=</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">Prevalence</span><span class="p">),</span> <span class="n">bins</span> <span class="o">=</span> <span class="m">60</span><span class="p">)</span>
</code></pre></div></p>
<p><img alt="" src="../../assets/images/Project3/step1_hist_prevalence.png" /></p>
<p>You can see that there are many species that apper only few times (on the left), and there are only few species that are present in all samples (on the right). We can also check which genera are the most prevalent:
<div class="highlight"><pre><span></span><code><span class="nf">head</span><span class="p">(</span><span class="n">prevalence_df</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">prevalence_df</span><span class="o">$</span><span class="n">Prevalence</span><span class="p">,</span><span class="n">decreasing</span> <span class="o">=</span> <span class="bp">T</span><span class="p">),])</span>
</code></pre></div>
Result:
<div class="highlight"><pre><span></span><code>                 <span class="n">Prevalence</span>            <span class="n">genus</span>
<span class="n">Blautia</span>           <span class="m">1.0000000</span>          <span class="n">Blautia</span>
<span class="n">Bacteroides</span>       <span class="m">1.0000000</span>      <span class="n">Bacteroides</span>
<span class="m">-1</span>                <span class="m">1.0000000</span>               <span class="m">-1</span>
<span class="n">Faecalibacterium</span>  <span class="m">0.9899194</span> <span class="n">Faecalibacterium</span>
<span class="n">Anaerostipes</span>      <span class="m">0.9879032</span>     <span class="n">Anaerostipes</span>
<span class="n">Fusicatenibacter</span>  <span class="m">0.9778226</span> <span class="n">Fusicatenibacter</span>
</code></pre></div>
It means that the genus <em>Blautia</em> and <em>Bacteroides</em> are present in all species. The <code>-1</code> represents unassigned reads (i.e. that they do not map to any known genus).</p>
<p>The least prevalent genera are:
<div class="highlight"><pre><span></span><code>                               <span class="n">Prevalence</span>                         <span class="n">genus</span>
<span class="n">Rosenbergiella</span>                <span class="m">0.002016129</span>                <span class="n">Rosenbergiella</span>
<span class="m">28-4</span>                          <span class="m">0.002016129</span>                          <span class="m">28-4</span>
<span class="n">Gallicola</span>                     <span class="m">0.002016129</span>                     <span class="n">Gallicola</span>
<span class="n">Sarcina</span>                       <span class="m">0.002016129</span>                       <span class="n">Sarcina</span>
<span class="n">Harryflintia</span>                  <span class="m">0.002016129</span>                  <span class="n">Harryflintia</span>
<span class="n">Paeniclostridium</span>              <span class="m">0.002016129</span>              <span class="n">Paeniclostridium</span>
</code></pre></div>
Which are present in only one sample.</p>
<p>You can see that the most prevalent species are typically genera that that should be present in all human guts. This type of quick exploration can also serve as a sanity check (is there something we should not be seeing at all?)</p>
</details>
</li>
<li>
<p>Is the relative abundance of the different genera normally distributed?
    <details>
    <summary markdown="span">Solution</summary></p>
<p>If we look at the distribution of all relative abundances with a simple histogram:</p>
<div class="highlight"><pre><span></span><code><span class="n">abundances_df</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">rel_abundances</span><span class="o">=</span><span class="nf">as.vector</span><span class="p">(</span><span class="n">rel_ab</span><span class="p">))</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">abundances_df</span><span class="p">)</span> <span class="o">+</span> <span class="nf">geom_histogram</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">rel_abundances</span><span class="p">),</span><span class="n">bins</span><span class="o">=</span><span class="m">100</span><span class="p">)</span> <span class="o">+</span> <span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;All relative abundances&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../../assets/images/Project3/step1_hist_rel_ab.png" /></p>
<p>We can clearly see that the relative abundances are not normally distributed. Maybe if we log transform the data, the result improve. Note that in order to log transform the data we need to add a small value (in this case <code>10^-4</code>) so that we don&rsquo;t have the problem of calculating the log of zero. Code:</p>
<div class="highlight"><pre><span></span><code><span class="n">log_rel_ab</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">rel_abundances</span><span class="o">=</span><span class="nf">log10</span><span class="p">(</span><span class="nf">as.vector</span><span class="p">(</span><span class="n">rel_ab</span><span class="p">)</span> <span class="o">+</span> <span class="m">10</span><span class="o">^</span><span class="m">-4</span><span class="p">))</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">log_rel_ab</span><span class="p">)</span> <span class="o">+</span> <span class="nf">geom_histogram</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">rel_abundances</span><span class="p">),</span><span class="n">bins</span><span class="o">=</span><span class="m">100</span><span class="p">)</span> <span class="o">+</span> <span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;All relative abundances (log10)&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">scale_y_log10</span><span class="p">()</span>
</code></pre></div>
<p><img alt="" src="../../assets/images/Project3/step1_hist_rel_ab_log.png" /></p>
<p>As you can see even when we log transform, the high number of zero makes the distribution not normal. </p>
<p>We can check also single genera. Here I select 3 genera: <em>Bacteroides</em> with prevalence of 1, <em>Akkermansia</em> with a prevalence of ~0.5 and <em>Harryflintia</em> with the lowest relative abundance. We can plot them with the following code:</p>
<div class="highlight"><pre><span></span><code><span class="n">df_genera</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">(</span>
    <span class="n">genus</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s">&quot;Bacteroides&quot;</span><span class="p">,</span><span class="nf">ncol</span><span class="p">(</span><span class="n">rel_ab</span><span class="p">)),</span><span class="nf">rep</span><span class="p">(</span><span class="s">&quot;Akkermansia&quot;</span><span class="p">,</span><span class="nf">ncol</span><span class="p">(</span><span class="n">rel_ab</span><span class="p">)),</span><span class="nf">rep</span><span class="p">(</span><span class="s">&quot;Harryflintia&quot;</span><span class="p">,</span><span class="nf">ncol</span><span class="p">(</span><span class="n">rel_ab</span><span class="p">))),</span>
    <span class="n">rel_ab</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="n">rel_ab</span><span class="p">[</span><span class="s">&quot;Bacteroides&quot;</span><span class="p">,],</span><span class="n">rel_ab</span><span class="p">[</span><span class="s">&quot;Akkermansia&quot;</span><span class="p">,],</span><span class="n">rel_ab</span><span class="p">[</span><span class="s">&quot;Harryflintia&quot;</span><span class="p">,])</span>
<span class="p">)</span>

<span class="nf">ggplot</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">df_genera</span><span class="p">)</span> <span class="o">+</span> <span class="nf">geom_histogram</span><span class="p">(</span><span class="n">mapping</span> <span class="o">=</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">rel_ab</span><span class="p">),</span> <span class="n">bins</span> <span class="o">=</span> <span class="m">60</span><span class="p">)</span> <span class="o">+</span> 
<span class="nf">facet_grid</span><span class="p">(</span><span class="n">.</span> <span class="o">~</span> <span class="n">genus</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../../assets/images/Project3/step_1_3genus_rel_ab.png" /></p>
<p>As you can see for <em>Harryflintia</em> there are almost only zeros (only one sample contain this genus). On the other hand <em>Bacteroides</em> can almost have a normal distribution (with a long tail on the right). While <em>Akkermansia</em> shows a tipical microbiome distribution plot with many samples where the measure is zero and then a tail with few samples where the relative abundance is higher.</p>
</details>
</li>
<li>
<p>How many zeros there are per sample and per genus?</p>
<p><details>
<summary markdown="span">Solution</summary></p>
<p>Taxonomic profiles are typically sparse because most species occur in frequently. Do we see this in our data as well? </p>
<p>How many 0s are present overall in the data? 
<div class="highlight"><pre><span></span><code><span class="c1"># we transform the relative abundance in a vector</span>
<span class="n">temp</span> <span class="o">=</span> <span class="nf">as.vector</span><span class="p">(</span><span class="n">rel_ab</span><span class="p">)</span>
<span class="c1"># and check the length (how many values there are)</span>
<span class="nf">length</span><span class="p">(</span><span class="n">temp</span><span class="p">)</span> <span class="c1"># 122016</span>
<span class="c1"># and how many values are zero:</span>
<span class="nf">sum</span><span class="p">(</span><span class="n">temp</span> <span class="o">==</span> <span class="m">0</span><span class="p">)</span> <span class="c1"># 94133</span>
</code></pre></div>
77% of the data are 0s (94133/122016)!</p>
<p>If we look at the percentage of 0s per sample:</p>
<div class="highlight"><pre><span></span><code><span class="nf">head</span><span class="p">(</span><span class="nf">data.frame</span><span class="p">(</span><span class="n">zeros_per_sample</span> <span class="o">=</span><span class="nf">colSums</span><span class="p">(</span><span class="n">rel_ab</span> <span class="o">==</span> <span class="m">0</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="nf">dim</span><span class="p">(</span><span class="n">rel_ab</span><span class="p">)[</span><span class="m">1</span><span class="p">])))</span>
</code></pre></div>
<p>Result:
<div class="highlight"><pre><span></span><code>          <span class="n">zeros_per_sample</span>
<span class="m">700002</span>_<span class="n">T0</span>        <span class="m">0.8536585</span>
<span class="m">700002</span>_<span class="n">T1</span>        <span class="m">0.7398374</span>
<span class="m">700002</span>_<span class="n">T2</span>        <span class="m">0.7520325</span>
<span class="m">700002</span>_<span class="n">T3</span>        <span class="m">0.7723577</span>
<span class="m">700004</span>_<span class="n">T0</span>        <span class="m">0.7723577</span>
<span class="m">700004</span>_<span class="n">T1</span>        <span class="m">0.7723577</span>
</code></pre></div></p>
<p>We can see that the sparsity is similar across all samples. Do you think that if we had samples from a different environment (like Soil for instance), we might see something different?</p>
<p>Percentage of 0s per genus is the same as <code>1 - prevalence</code>:
<div class="highlight"><pre><span></span><code><span class="nf">head</span><span class="p">(</span><span class="nf">data.frame</span><span class="p">(</span><span class="n">zeros_per_sample</span> <span class="o">=</span><span class="nf">rowSums</span><span class="p">(</span><span class="n">rel_ab</span> <span class="o">==</span> <span class="m">0</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="nf">dim</span><span class="p">(</span><span class="n">rel_ab</span><span class="p">)[</span><span class="m">2</span><span class="p">])))</span>
</code></pre></div>
Result:
<div class="highlight"><pre><span></span><code>                 <span class="n">zeros_per_sample</span>
<span class="n">Blautia</span>                <span class="m">0.00000000</span>
<span class="n">Bacteroides</span>            <span class="m">0.00000000</span>
<span class="n">Agathobacter</span>           <span class="m">0.04435484</span>
<span class="n">Faecalibacterium</span>       <span class="m">0.01008065</span>
<span class="n">Bifidobacterium</span>        <span class="m">0.09072581</span>
<span class="n">Fusicatenibacter</span>       <span class="m">0.02217742</span>
</code></pre></div></p>
<p>Again, we note that some genera are more prevalent than others.</p>
</details>
</li>
<li>
<p>How much variability there is within Subject (check the <code>metadata</code> table), compare to between subjects? Or from another perspective, how stable it is the human gut microbiome?</p>
<p><details>
<summary markdown="span">Solution 1</summary></p>
<p>There are different ways to explore this problem. We can try to calculate the distance between all possible samples and then compare the distances of samples that come from the same subject and the distance that come from different subjects.
For example, you take sample 1 (<code>700002_T0</code>) and sample 2 (<code>700002_T1</code>), and since they are from the same subject, we will use the distance between these two samples as an example of <em>within-subject</em> distance.</p>
<p>Code:
<div class="highlight"><pre><span></span><code><span class="n">rel_ab</span> <span class="o">=</span> <span class="nf">prop.table</span><span class="p">(</span><span class="n">tax_profile</span><span class="p">,</span><span class="m">2</span><span class="p">)</span>
<span class="n">log_rel_ab</span> <span class="o">=</span> <span class="nf">log10</span><span class="p">(</span><span class="n">rel_ab</span><span class="o">+</span> <span class="m">10</span><span class="o">^</span><span class="m">-4</span><span class="p">)</span>

<span class="n">n_samples</span> <span class="o">=</span> <span class="nf">ncol</span><span class="p">(</span><span class="n">log_rel_ab</span><span class="p">)</span>

<span class="c1"># where to save the result</span>
<span class="n">distance</span> <span class="o">=</span> <span class="nf">c</span><span class="p">()</span>
<span class="n">type</span> <span class="o">=</span> <span class="nf">c</span><span class="p">()</span>

<span class="c1"># we go through all the pair of samples and we calculate the distance</span>
<span class="nf">for</span><span class="p">(</span><span class="n">i</span> <span class="n">in</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span> <span class="p">(</span><span class="n">n_samples</span><span class="m">-1</span><span class="p">)</span> <span class="p">)){</span>
  <span class="nf">for</span><span class="p">(</span><span class="n">j</span> <span class="n">in</span> <span class="nf">c</span><span class="p">(</span> <span class="p">(</span><span class="n">i</span><span class="m">+1</span><span class="p">)</span> <span class="o">:</span> <span class="n">n_samples</span><span class="p">)){</span>
    <span class="c1"># name of the samples</span>
    <span class="n">s_i</span> <span class="o">=</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">log_rel_ab</span><span class="p">)[</span><span class="n">i</span><span class="p">]</span>
    <span class="n">s_j</span> <span class="o">=</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">log_rel_ab</span><span class="p">)[</span><span class="n">j</span><span class="p">]</span>
    <span class="c1"># profiles</span>
    <span class="n">profile_i</span> <span class="o">=</span> <span class="n">log_rel_ab</span><span class="p">[,</span><span class="n">i</span><span class="p">]</span>
    <span class="n">profile_j</span> <span class="o">=</span> <span class="n">log_rel_ab</span><span class="p">[,</span><span class="n">j</span><span class="p">]</span>
    <span class="c1"># calculate the distance as sum of the absolute distance</span>
    <span class="n">d</span> <span class="o">=</span> <span class="nf">sum</span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="n">profile_i</span> <span class="o">-</span> <span class="n">profile_j</span><span class="p">))</span>
    <span class="c1"># add the values</span>
    <span class="n">distance</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="n">distance</span><span class="p">,</span><span class="n">d</span><span class="p">)</span>
    <span class="c1"># add correct type</span>
    <span class="nf">if</span><span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="n">s_i</span><span class="p">,</span><span class="s">&quot;Subject&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">metadata</span><span class="p">[</span><span class="n">s_j</span><span class="p">,</span><span class="s">&quot;Subject&quot;</span><span class="p">]){</span>
      <span class="n">type</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="n">type</span><span class="p">,</span><span class="s">&quot;Within distance&quot;</span><span class="p">)</span>
    <span class="p">}</span><span class="n">else</span><span class="p">{</span>
      <span class="n">type</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="n">type</span><span class="p">,</span><span class="s">&quot;Between distance&quot;</span><span class="p">)</span>
    <span class="p">}</span>
  <span class="p">}</span>
<span class="p">}</span>

<span class="n">df</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">dist</span> <span class="o">=</span> <span class="n">distance</span><span class="p">,</span><span class="n">type</span> <span class="o">=</span> <span class="n">type</span><span class="p">)</span>

<span class="nf">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="nf">aes</span><span class="p">(</span><span class="n">type</span><span class="p">,</span><span class="n">dist</span><span class="p">))</span> <span class="o">+</span> <span class="nf">geom_boxplot</span><span class="p">()</span>
</code></pre></div></p>
<p><img alt="" src="../../assets/images/Project3/step_1_distances.png" /></p>
<p>It is quite clear that they are different, but we can also test it:
<div class="highlight"><pre><span></span><code><span class="nf">wilcox.test</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">$</span><span class="n">type</span> <span class="o">==</span> <span class="s">&quot;Within distance&quot;</span><span class="p">,]</span><span class="o">$</span><span class="n">dist</span><span class="p">,</span>
            <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">$</span><span class="n">type</span> <span class="o">==</span> <span class="s">&quot;Between distance&quot;</span><span class="p">,]</span><span class="o">$</span><span class="n">dist</span><span class="p">)</span>
</code></pre></div></p>
<p>Result:
<div class="highlight"><pre><span></span><code><span class="n">Wilcoxon</span> <span class="n">rank</span> <span class="n">sum</span> <span class="n">test</span> <span class="n">with</span> <span class="n">continuity</span> <span class="n">correction</span>

<span class="n">data</span><span class="o">:</span>  <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">$</span><span class="n">type</span> <span class="o">==</span> <span class="s">&quot;Within distance&quot;</span><span class="p">,</span> <span class="p">]</span><span class="o">$</span><span class="n">dist</span> <span class="n">and</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">$</span><span class="n">type</span> <span class="o">==</span> <span class="s">&quot;Between distance&quot;</span><span class="p">,</span> <span class="p">]</span><span class="o">$</span><span class="n">dist</span>
<span class="n">W</span> <span class="o">=</span> <span class="m">2844084</span><span class="p">,</span> <span class="n">p</span><span class="o">-</span><span class="n">value</span> <span class="o">&lt;</span> <span class="m">2.2e-16</span>
<span class="n">alternative</span> <span class="n">hypothesis</span><span class="o">:</span> <span class="n">true</span> <span class="n">location</span> <span class="n">shift</span> <span class="n">is</span> <span class="n">not</span> <span class="n">equal</span> <span class="n">to</span> <span class="m">0</span>
</code></pre></div></p>
<p>This means that samples from the same subject have a low distance, or in other words they are similar to each other. And they are much more similar compared to other subjects, even after 1 year (the last time point is 50 weeks). From this we understand two things:
- First, the human microbiome is stable over time
- Second, there is a great variability between subjects</p>
</details>
<p><details>
<summary markdown="span">Solution 2</summary></p>
<p>We can do a PCA plot:
<div class="highlight"><pre><span></span><code><span class="n">rel_ab</span> <span class="o">=</span> <span class="nf">prop.table</span><span class="p">(</span><span class="n">tax_profile</span><span class="p">,</span><span class="m">2</span><span class="p">)</span>
<span class="n">log_rel_ab</span> <span class="o">=</span> <span class="nf">log10</span><span class="p">(</span><span class="n">rel_ab</span><span class="o">+</span> <span class="m">10</span><span class="o">^</span><span class="m">-4</span><span class="p">)</span>

<span class="c1"># remove zero rows</span>
<span class="n">log_rel_ab</span> <span class="o">=</span> <span class="n">log_rel_ab</span><span class="p">[</span><span class="nf">rowSums</span><span class="p">(</span><span class="n">rel_ab</span><span class="p">)</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">,]</span>

<span class="n">pc</span> <span class="o">&lt;-</span> <span class="nf">prcomp</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">log_rel_ab</span><span class="p">),</span>
             <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
             <span class="n">scale.</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>

<span class="n">df</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">(</span>
  <span class="n">pc1</span> <span class="o">=</span> <span class="n">pc</span><span class="o">$</span><span class="n">x</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span>
  <span class="n">pc2</span> <span class="o">=</span> <span class="n">pc</span><span class="o">$</span><span class="n">x</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span>
  <span class="n">Subject</span> <span class="o">=</span> <span class="nf">as.factor</span><span class="p">(</span><span class="n">metadata</span><span class="p">[</span><span class="nf">rownames</span><span class="p">(</span><span class="n">pc</span><span class="o">$</span><span class="n">x</span><span class="p">),</span><span class="s">&quot;Subject&quot;</span><span class="p">]),</span>
  <span class="n">Timepoint</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="nf">rownames</span><span class="p">(</span><span class="n">pc</span><span class="o">$</span><span class="n">x</span><span class="p">),</span><span class="s">&quot;Timepoint&quot;</span><span class="p">],</span>
  <span class="n">Sex</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="nf">rownames</span><span class="p">(</span><span class="n">pc</span><span class="o">$</span><span class="n">x</span><span class="p">),</span><span class="s">&quot;Sex&quot;</span><span class="p">]</span>
<span class="p">)</span>
</code></pre></div></p>
<p>And plot the result:
<div class="highlight"><pre><span></span><code><span class="nf">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">pc1</span><span class="p">,</span><span class="n">y</span> <span class="o">=</span> <span class="n">pc2</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">Sex</span><span class="p">))</span> <span class="o">+</span> <span class="nf">geom_point</span><span class="p">()</span>
</code></pre></div></p>
<p><img alt="" src="../../assets/images/Project3/step_1_pca_sex.png" /></p>
<div class="highlight"><pre><span></span><code><span class="nf">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">pc1</span><span class="p">,</span><span class="n">y</span> <span class="o">=</span> <span class="n">pc2</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">Timepoint</span><span class="p">))</span> <span class="o">+</span> <span class="nf">geom_point</span><span class="p">()</span>
</code></pre></div>
<p><img alt="" src="../../assets/images/Project3/step_1_pca_timepoint.png" /></p>
<div class="highlight"><pre><span></span><code><span class="nf">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="n">pc1</span><span class="p">,</span><span class="n">y</span> <span class="o">=</span> <span class="n">pc2</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">Subject</span><span class="p">))</span> <span class="o">+</span> <span class="nf">geom_point</span><span class="p">()</span> <span class="o">+</span>
<span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span> <span class="o">=</span> <span class="s">&quot;none&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../../assets/images/Project3/step_1_pca_subject.png" /></p>
<p>There is no particular grouping by sex or by timepoint (as expected). It also seems that the plot based on subject (third plot where each color is a subject) is random. But, we can actually see some structure, in particular outside of the conglomerate of points at the center. </p>
</details>
</li>
</ul>

              
            </article>
            
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../Step-1/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Step 1" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Step 1
            </div>
          </div>
        </a>
      
      
        
        <a href="../Step-2/" class="md-footer__link md-footer__link--next" aria-label="Next: Step 2" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Step 2
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.b97dbffb.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.6c7ad80a.min.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>